### PR TITLE
Preserve storage size when generating functional tensor

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128546

Original PR from shazqadeer at
https://github.com/pytorch/pytorch/pull/128141

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @chauhang

### CODE REVIEW

**Comment by ezyang** [comment]
This is good to go. @bdhirsh you might be the right person for the review here.

File: torch/distributed/_tensor/_sharding_prop.py [hunk 1]
```diff
@@ -104,7 +105,7 @@ def _propagate_tensor_meta(
 
         # NOTE: We must call the tracing in fake tensor mode so that it
         # avoids materializing memory
-        with FakeTensorMode():
+        with detect_fake_mode() or FakeTensorMode():
```
Comment (bdhirsh) [inline torch/distributed/_tensor/_sharding_prop.py:108]: unrelated change? although seems like a strict improvement

**Comment by bdhirsh** [comment]
Sorry for the late review. I don't see any tests in this PR but I'm wondering if it is obviated by this change last month? https://github.com/pytorch/pytorch/pull/132524

The main annoyance/problem was that the "storage preservation" logic would use `aten.set_()` to slam storages from the input to the output, but this ran `aten.set_`'s shape checks that could cause extra guards. I added a helper that just does the storage slamming unsafely, which I think should prevent the need to suppress in the first place (and also probably improve compile times a bit, although I wasn't able to get great measurements)

**Comment by ezyang** [comment]
ok, we can kibosh this, I don't even remember what we needed it for
